adapters/beachfront/beachfront.go

```diff
@@ -218,40 +216,82 @@ func preprocess(request *openrtb.BidRequest) (beachfrontReqs beachfrontRequests,
 	}
 
 	if len(bannerImps) > 0 {
-		request.Imp = bannerImps
+		// request.Imp = bannerImps		// ... why was I doing this????
 		beachfrontReqs.Banner, errs = getBannerRequest(request)
 	}
 
+	/* We have video imps, so organize them into nurl imps and adm imps. The nurl imps will be
+	sent sequentially and the adm imps in sequential / parallel.
+	*/
 	if len(videoImps) > 0 {
-		var videoErrs []error
-		var videoList []beachfrontVideoRequest
+		admRequests := make(map[string]videoRequest, 0)
```

Comment:
We might want to give go a hint to the map size with `admRequests := make(map[string]videoRequest, len(videoImps))` if the imp `AppId`s tend to be unique. If it is expected that the imps usually share the same `AppId`, then the current line should be fine.

Reply:
I was concerned about making them too long, with null members at the end, but since they don't get returned, and I never call len() on any of them or iterate through them, just pull from them to append members to the slice I'm building for return, I think that's not a problem. I've identified a couple of other concerns similar to this. I'm not sure about the relative expense in one particular situation, though.  At 187 in the current commit: 

```
func preprocess(request *openrtb.BidRequest) (beachfrontReqs requests, errs []error) {
	var videoImps = make([]openrtb.Imp, 0)
	var bannerImps = make([]openrtb.Imp, 0)
```

At this point I don't know if I have any videoImps or bannerImps, so I started them at 0, then as I find Imps, I append to them. Considering that they are pretty much throw away slices, I would be better off for efficiency, as above, making some extra then assigning instead of appending, like this:

```
bannerImps = make([]openrtb.Imp, len(request.Imp))
videoImps = make([]openrtb.Imp, len(request.Imp))

```
What I'm not sure about is maybe I would be better off defining them like this:

```
	var videoImps []openrtb.Imp
	var bannerImps []openrtb.Imp
```

And only make the slice when I hit my first imp of that type, and make it len(request.Imp), something like this:

```
		if request.Imp[i].Video != nil {
			if noVideoImpsYet {
				noVideoImpsYet = false
				videoImps = make([]openrtb.Imp, 1)
			}
```
But does checking those couple of "if"s and reassigning the flags add significant overhead that would indicate I should just make them len(request.Imp) long in the first place?

Reply:
You want to take advantage of the third parameter in `make()`. If you have:
```
bannerImps = make([]openrtb.Imp, 0, len(request.Imp))
videoImps = make([]openrtb.Imp, 0, len(request.Imp))
```
you will have zero length slices where you will have to use append to add any elements to the slice. So the size of the slice is 0, but the capacity of the slice is `len(request.Imp)`. In Go, each slice points to an existing, allocated array. The above makes create the slices pointing to an array that has `len(request.Imp)` items of memory already allocated. As you append items, the slice will expand into the unused elements of the array that are sitting there waiting to be used. If you try to append past the capacity, then Go will allocate a new, larger array, copy the contents of the original to the larger, and then put the new item into the next element. This is why we like to avoid these sorts of reallocations when we have some idea what size could be.